src/plugins/initiatives/discourse.test.js

```diff
@@ -0,0 +1,271 @@
+        timestampMs: 1571498171951,
+        authorUsername: "TestUser",
+      };
+      const firstPost = {
+        id: 432,
+        topicId: 123,
+        indexWithinTopic: 1,
+        replyToPostIndex: null,
+        timestampMs: 1571498171951,
+        authorUsername: "TestUser",
+        cooked: `
+            <h1>Example initiative</h1>
+            <h2>Status: Testing</h2>
+            <h2>Champion<a href="https://foo.bar/t/dont-include/10"><sup>?</sup></a>:</h2>
+            <p>
+              <a class="mention" href="/u/ChampUser">@ChampUser</a>
+            </p>
+            <h2>Dependencies:</h2>
+            <ul>
+              <li><a href="https://foo.bar/t/dependency/1">Thing we need</a></li>
+              <li><a href="https://foo.bar/t/dependency/2">Thing we need</a></li>
+              <li><a href="https://foo.bar/t/dependency/3">Thing we need</a></li>
+            </ul>
+            <h2>References:</h2>
+            <ul>
+              <li><a href="https://foo.bar/t/reference/4">Some reference</a></li>
+              <li><a href="https://foo.bar/t/reference/5/2">Some reference</a></li>
+              <li><a href="https://foo.bar/t/reference/6/4">Some reference</a></li>
+            </ul>
+            <h2>Contributions:</h2>
+            <ul>
+              <li><a href="https://foo.bar/t/contribution/7">Some contribution</a></li>
+              <li><a href="https://foo.bar/t/contribution/8/2">Some contribution</a></li>
+              <li><a href="https://github.com/sourcecred/sourcecred/pull/1416">Some contribution</a></li>
+            </ul>
+          `,
+      };
+
+      // When
+      const initiative = await initiativeFromDiscourseTracker(
+        serverUrl,
+        topic,
+        firstPost
+      );
+
+      // Then
+      expect(initiative.completed).toEqual(false);
+      expect(initiative).toMatchSnapshot();
+    });
+
+    it("handles a completed status", async () => {
+      // Given
+      const serverUrl = "https://foo.bar";
+      const topic = {
+        id: 123,
+        title: "Example initiative",
+        timestampMs: 1571498171951,
+        authorUsername: "TestUser",
+      };
+      const firstPost = {
+        id: 432,
+        topicId: 123,
+        indexWithinTopic: 1,
+        replyToPostIndex: null,
+        timestampMs: 1571498171951,
+        authorUsername: "TestUser",
+        cooked: `
+            <h1>Example initiative</h1>
+            <h2>Status: Complete</h2>
+            <h2>Champion:</h2>
+            <h2>Dependencies:</h2>
+            <h2>References:</h2>
+            <h2>Contributions:</h2>
+          `,
+      };
+
+      // When
+      const initiative = await initiativeFromDiscourseTracker(
+        serverUrl,
+        topic,
+        firstPost
+      );
+
+      // Then
+      expect(initiative.completed).toEqual(true);
+      expect(initiative).toMatchSnapshot();
+    });
+
+    it("considers blank status incomplete", async () => {
+      // Given
```

Comment:
The `// Given`, `// When` and `// Then` pattern is new to me. it seems a little redundant with just reading the code, can you explain what the motivation is?

Reply:
It's a convention I picked up from some BDD promotion blogs. It encourages you to separate input parameters (given) from the code being tested (when) and your assertion code (then), it's intended to make tests faster to read.

An example of where this influenced how I wrote tests.
Jest documentation has [example code](https://jestjs.io/docs/en/expect.html#tothrowerror) for toThrow like this:
```js
expect(() => {
  drinkFlavor('octopus');
}).toThrow();
```

Which conflates all of given, when and then, so I would have written it like:

```js

// Given
const flavor = 'octopus';

// When
const fn = () => drinkFlavor(flavor);

// Then
expect(fn).toThrow();
```

Examples of this being used with promise rejection:

```js
// When
const initiativeP = initiativeFromDiscourseTracker(
  serverUrl,
  topic,
  firstPost
);

// Then
await expect(initiativeP).rejects.toThrow(
  'Missing or malformed headers ["status","champions","dependencies","references","contributions"] for initiative topic "Example initiative" (123)'
);
```